File: tools/qaimodelbuilder/src/qai/chat/domain/agent_template.py
# ---------------------------------------------------------------------
# Copyright (c) 2026 Qualcomm Technologies, Inc. and/or its subsidiaries.
# SPDX-License-Identifier: BSD-3-Clause
# ---------------------------------------------------------------------

"""``AgentTemplate`` aggregate for the chat bounded context.

An :class:`AgentTemplate` is a *named, reusable definition of a SINGLE
discussion role* — an "agent" such as ``资深架构师`` / ``全栈开发`` /
``严谨测试`` — that a user can preview and import into any conversation (or pull
into a team).  It is the smallest reusable unit in the three-tier template
system (design §27: single role → team → mode), orthogonal to the team-level
:class:`~qai.chat.domain.roster_template.RosterTemplate` and the
mode-level ``ModeTemplate``.

Relationship to :class:`~qai.chat.domain.participant.Participant`
----------------------------------------------------------------
A ``Participant`` is strictly conversation-scoped (one row per speaker, bound
to a ``conversation_id``).  An ``AgentTemplate`` is the orthogonal,
*conversation-independent* library entry: it holds one role *definition*
(display name + model + persona + config) WITHOUT any conversation binding.
"Applying" a template to a conversation instantiates exactly one
``kind=named_agent`` ``Participant``.

Mutability model
----------------
The aggregate is mutable (``name`` / ``description`` / ``display_name`` /
``model_id`` / ``persona`` / ``config`` may change via mutators that bump
``updated_at``).  Like :class:`Participant`, mutators emit no platform
side-effects (domain stays pure).  ``is_builtin`` templates are factory-seeded
presets and are treated as read-only by the application layer.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any

from qai.chat.domain.ids import AgentTemplateId
from qai.platform.time import ensure_aware_utc

_MAX_NAME_LENGTH: int = 256
_MAX_DESCRIPTION_LENGTH: int = 2000
_MAX_DISPLAY_NAME_LENGTH: int = 256


def _validate_config(config: dict[str, Any] | None) -> dict[str, Any] | None:
    """Validate + defensively copy a config blob.

    Recognises the same keys as ``Participant.config`` (``allowed_tools`` /
    ``color``) so the agent can be instantiated into a participant 1:1.
    """
    if config is None:
        return None
    if not isinstance(config, dict):
        raise TypeError("AgentTemplate.config must be a dict or None")
    allowed = config.get("allowed_tools")
    if allowed is not None and (
        not isinstance(allowed, (list, tuple))
        or not all(isinstance(tool, str) for tool in allowed)
    ):
        raise TypeError(
            "AgentTemplate.config['allowed_tools'] must be a list of str "
            "when present",
        )
    color = config.get("color")
    if color is not None and (
        isinstance(color, bool) or not isinstance(color, (str, int))
    ):
        raise TypeError(
            "AgentTemplate.config['color'] must be a str, int (palette index) "
            "or None",
        )
    copied: dict[str, Any] = dict(config)
    if allowed is not None:
        copied["allowed_tools"] = list(allowed)
    return copied


@dataclass(slots=True, kw_only=True)
class AgentTemplate:
    """A named, reusable definition of a single discussion role.

    Construction goes through :meth:`create` to set ``created_at`` /
    ``updated_at`` consistently; loading from persistence uses the regular
    constructor.
    """

    id: AgentTemplateId
    name: str = ""
    description: str = ""
    display_name: str = ""
    model_id: str | None = None
    persona: str | None = None
    config: dict[str, Any] | None = None
    is_builtin: bool = False
    #: When this template was cloned from another (e.g. a built-in preset), the
    #: SOURCE template's id; ``None`` for originals. Tail-appended optional field
    #: (§3.1 additive). "Editing a built-in preset" is modelled as cloning a
    #: user copy (``is_builtin=False`` + ``cloned_from_id=preset.id``) and then
    #: editing the copy — the original stays read-only and untouched. ``reset``
    #: restores a copy's business fields from ``cloned_from_id`` in place.
    cloned_from_id: str | None = None
    #: Optional per-locale i18n maps for built-in presets (migration 056).
    #: Each is ``{"en": "...", "zh-CN": "...", "zh-TW": "..."}`` loaded from the
    #: matching ``*_i18n_json`` column; ``None`` for custom (``is_builtin=0``)
    #: templates and pre-056 rows — which then always render their canonical
    #: single-language field (name / description / display_name / persona) as the
    #: fallback (AGENTS.md §8 forward-compatibility). Tail-appended optional
    #: fields (§3.1 additive) so old constructors / old data are unaffected.
    name_i18n: dict[str, str] | None = None
    description_i18n: dict[str, str] | None = None
    display_name_i18n: dict[str, str] | None = None
    persona_i18n: dict[str, str] | None = None
    created_at: datetime
    updated_at: datetime

    def __post_init__(self) -> None:
        if not isinstance(self.id, AgentTemplateId):
            raise TypeError(
                "AgentTemplate.id must be AgentTemplateId, got "
                f"{type(self.id).__name__}",
            )
        if not isinstance(self.name, str):
            raise TypeError("AgentTemplate.name must be a str")
        if len(self.name) > _MAX_NAME_LENGTH:
            raise ValueError(
                f"AgentTemplate.name must be <= {_MAX_NAME_LENGTH} chars",
            )
        if not isinstance(self.description, str):
            raise TypeError("AgentTemplate.description must be a str")
        if len(self.description) > _MAX_DESCRIPTION_LENGTH:
            raise ValueError(
                "AgentTemplate.description must be "
                f"<= {_MAX_DESCRIPTION_LENGTH} chars",
            )
        if not isinstance(self.display_name, str):
            raise TypeError("AgentTemplate.display_name must be a str")
        if len(self.display_name) > _MAX_DISPLAY_NAME_LENGTH:
            raise ValueError(
                "AgentTemplate.display_name must be "
                f"<= {_MAX_DISPLAY_NAME_LENGTH} chars",
            )
        if self.model_id is not None and not isinstance(self.model_id, str):
            raise TypeError("AgentTemplate.model_id must be a str or None")
        if self.persona is not None and not isinstance(self.persona, str):
            raise TypeError("AgentTemplate.persona must be a str or None")
        self.config = _validate_config(self.config)
        if not isinstance(self.is_builtin, bool):
            raise TypeError("AgentTemplate.is_builtin must be a bool")
        if self.cloned_from_id is not None and not isinstance(
            self.cloned_from_id, str
        ):
            raise TypeError("AgentTemplate.cloned_from_id must be a str or None")
        ensure_aware_utc(self.created_at)
        ensure_aware_utc(self.updated_at)

    # ------------------------------------------------------------------
    # Factories
    # ------------------------------------------------------------------
    @classmethod
    def create(
        cls,
        *,
        template_id: AgentTemplateId,
        now: datetime,
        name: str = "",
        description: str = "",
        display_name: str = "",
        model_id: str | None = None,
        persona: str | None = None,
        config: dict[str, Any] | None = None,
        is_builtin: bool = False,
        cloned_from_id: str | None = None,
    ) -> AgentTemplate:
        """Construct a brand-new agent template."""
        ts = ensure_aware_utc(now)
        return cls(
            id=template_id,
            name=name,
            description=description,
            display_name=display_name,
            model_id=model_id,
            persona=persona,
            config=config,
            is_builtin=is_builtin,
            cloned_from_id=cloned_from_id,
            created_at=ts,
            updated_at=ts,
        )

    # ------------------------------------------------------------------
    # Mutators
    # ------------------------------------------------------------------
    def rename(self, name: str, *, now: datetime) -> None:
        """Set a new template name and bump ``updated_at``."""
        if not isinstance(name, str):
            raise TypeError("name must be a str")
        if len(name) > _MAX_NAME_LENGTH:
            raise ValueError(f"name must be <= {_MAX_NAME_LENGTH} chars")
        self.name = name
        self.updated_at = ensure_aware_utc(now)

    def set_description(self, description: str, *, now: datetime) -> None:
        """Replace the description and bump ``updated_at``."""
        if not isinstance(description, str):
            raise TypeError("description must be a str")
        if len(description) > _MAX_DESCRIPTION_LENGTH:
            raise ValueError(
                f"description must be <= {_MAX_DESCRIPTION_LENGTH} chars",
            )
        self.description = description
        self.updated_at = ensure_aware_utc(now)

    def set_role(
        self,
        *,
        now: datetime,
        display_name: str,
        model_id: str | None,
        persona: str | None,
        config: dict[str, Any] | None,
    ) -> None:
        """Replace the role definition (display_name / model / persona / config).

        Bumps ``updated_at``.  Validation mirrors ``__post_init__``.
        """
        if not isinstance(display_name, str):
            raise TypeError("display_name must be a str")
        if len(display_name) > _MAX_DISPLAY_NAME_LENGTH:
            raise ValueError(
                f"display_name must be <= {_MAX_DISPLAY_NAME_LENGTH} chars",
            )
        if model_id is not None and not isinstance(model_id, str):
            raise TypeError("model_id must be a str or None")
        if persona is not None and not isinstance(persona, str):
            raise TypeError("persona must be a str or None")
        self.display_name = display_name
        self.model_id = model_id
        self.persona = persona
        self.config = _validate_config(config)
        self.updated_at = ensure_aware_utc(now)


__all__ = [
    "AgentTemplate",
]

File: tools/qaimodelbuilder/frontend/src/stores/agentTemplate.ts
// ---------------------------------------------------------------------
// Copyright (c) 2026 Qualcomm Technologies, Inc. and/or its subsidiaries.
// SPDX-License-Identifier: BSD-3-Clause
// ---------------------------------------------------------------------

/**
 * Agent-template store — reusable single-role "agents".
 *
 * Thin CRUD client for the conversation-INDEPENDENT agent-template library
 * (the smallest reusable unit in the three-tier template system, §27):
 *
 *   GET    /api/chat/agent-templates
 *   POST   /api/chat/agent-templates
 *   PATCH  /api/chat/agent-templates/{id}
 *   DELETE /api/chat/agent-templates/{id}
 *   POST   /api/chat/agent-templates/{id}/apply   (instantiate into a conversation)
 *
 * An agent template is a named definition of a SINGLE role (display_name /
 * model_id / persona / config) a user can preview + import into any
 * conversation, so a frequently-used role need not be re-typed. Built-in
 * presets (``is_builtin``) are factory-seeded and read-only.
 *
 * PURE V2 enhancement (V1 has no multi-Agent discussion). The store owns the
 * wire shapes + HTTP only; it maps snake_case wire ↔ camelCase view models.
 */
import { defineStore } from "pinia";
import { ref } from "vue";
import { apiJson } from "@/api";

// ---------------------------------------------------------------------------
// Wire shapes (snake_case — backend contract)
// ---------------------------------------------------------------------------

interface AgentConfigWire {
  allowed_tools?: string[] | null;
  enabled_skills?: string[] | null;
  color?: number | string | null;
}

interface AgentTemplateWire {
  id: string;
  name: string;
  description: string;
  display_name: string;
  model_id?: string | null;
  persona?: string | null;
  config?: AgentConfigWire | null;
  is_builtin: boolean;
  cloned_from_id?: string | null;
  created_at: string;
  updated_at: string;
  /** Per-locale i18n maps for built-in presets (migration 056); null/absent
   *  for custom rows → fall back to the single-language fields above. */
  name_i18n?: Record<string, string> | null;
  description_i18n?: Record<string, string> | null;
  display_name_i18n?: Record<string, string> | null;
  persona_i18n?: Record<string, string> | null;
}

interface AgentTemplateListWire {
  items?: AgentTemplateWire[] | null;
}

interface ApplyResponseWire {
  conversation_id: string;
  participant_id: string;
}

// ---------------------------------------------------------------------------
// View models (camelCase)
// ---------------------------------------------------------------------------

export interface AgentTemplateView {
  id: string;
  name: string;
  description: string;
  displayName: string;
  modelId?: string;
  persona?: string;
  allowedTools: string[];
  enabledSkills: string[];
  color?: number;
  isBuiltin: boolean;
  /** Source template id when this is a clone (esp. a clone of a factory preset);
   *  "" / undefined = not a clone. Reset is only meaningful when set. */
  clonedFromId?: string;
  /** Per-locale i18n maps for built-in presets; undefined for custom rows.
   *  Consumed by useTemplateI18n at the display layer to localise built-in
   *  text without duplicating translations into the frontend locale files. */
  nameI18n?: Record<string, string>;
  descriptionI18n?: Record<string, string>;
  displayNameI18n?: Record<string, string>;
  personaI18n?: Record<string, string>;
}

/** Body for create / update (id is route/response only). */
export interface AgentTemplateInput {
  name: string;
  description: string;
  displayName: string;
  modelId?: string;
  persona?: string;
  allowedTools: string[];
  enabledSkills: string[];
  color?: number;
}

// ---------------------------------------------------------------------------
// Wire ↔ view-model mappers
// ---------------------------------------------------------------------------

function wireToTemplate(w: AgentTemplateWire): AgentTemplateView {
  return {
    id: w.id,
    name: w.name,
    description: w.description,
    displayName: w.display_name,
    ...(w.model_id != null && w.model_id !== "" ? { modelId: w.model_id } : {}),
    ...(w.persona != null && w.persona !== "" ? { persona: w.persona } : {}),
    allowedTools: Array.isArray(w.config?.allowed_tools)
      ? [...(w.config?.allowed_tools as string[])]
      : [],
    enabledSkills: Array.isArray(w.config?.enabled_skills)
      ? [...(w.config?.enabled_skills as string[])]
      : [],
    ...(typeof w.config?.color === "number" ? { color: w.config?.color } : {}),
    isBuiltin: w.is_builtin === true,
    ...(w.cloned_from_id != null && w.cloned_from_id !== ""
      ? { clonedFromId: w.cloned_from_id }
      : {}),
    ...(w.name_i18n != null ? { nameI18n: w.name_i18n } : {}),
    ...(w.description_i18n != null
      ? { descriptionI18n: w.description_i18n }
      : {}),
    ...(w.display_name_i18n != null
      ? { displayNameI18n: w.display_name_i18n }
      : {}),
    ...(w.persona_i18n != null ? { personaI18n: w.persona_i18n } : {}),
  };
}

function inputToWire(input: AgentTemplateInput): Record<string, unknown> {
  return {
    name: input.name,
    description: input.description,
    display_name: input.displayName,
    ...(input.modelId !== undefined && input.modelId !== ""
      ? { model_id: input.modelId }
      : {}),
    ...(input.persona !== undefined && input.persona !== ""
      ? { persona: input.persona }
      : {}),
    config: {
      allowed_tools: input.allowedTools,
      enabled_skills: input.enabledSkills,
      ...(input.color !== undefined ? { color: input.color } : {}),
    },
  };
}

// ---------------------------------------------------------------------------
// Store
// ---------------------------------------------------------------------------

export const useAgentTemplateStore = defineStore("agentTemplate", () => {
  /** Cached library (built-ins first, then user saved). */
  const templates = ref<AgentTemplateView[]>([]);
  const loaded = ref(false);

  async function fetchAll(): Promise<AgentTemplateView[]> {
    const resp = await apiJson<AgentTemplateListWire>(
      "GET",
      "/api/chat/agent-templates",
    );
    const items = Array.isArray(resp?.items) ? resp.items : [];
    templates.value = items.map(wireToTemplate);
    loaded.value = true;
    return templates.value;
  }

  async function create(input: AgentTemplateInput): Promise<AgentTemplateView> {
    const w = await apiJson<AgentTemplateWire>(
      "POST",
      "/api/chat/agent-templates",
      inputToWire(input),
    );
    const view = wireToTemplate(w);
    templates.value = [...templates.value, view];
    return view;
  }

  async function update(
    id: string,
    input: AgentTemplateInput,
  ): Promise<AgentTemplateView> {
    const w = await apiJson<AgentTemplateWire>(
      "PATCH",
      `/api/chat/agent-templates/${encodeURIComponent(id)}`,
      inputToWire(input),
    );
    const view = wireToTemplate(w);
    templates.value = templates.value.map((tpl) => (tpl.id === id ? view : tpl));
    return view;
  }

  async function remove(id: string): Promise<void> {
    await apiJson("DELETE", `/api/chat/agent-templates/${encodeURIComponent(id)}`);
    templates.value = templates.value.filter((tpl) => tpl.id !== id);
  }

  /** Apply (import) an agent template into a conversation as one named agent.
   *  Returns the created participant id. */
  async function applyToConversation(
    id: string,
    conversationId: string,
  ): Promise<string> {
    const resp = await apiJson<ApplyResponseWire, { conversation_id: string }>(
      "POST",
      `/api/chat/agent-templates/${encodeURIComponent(id)}/apply`,
      { conversation_id: conversationId },
    );
    return resp.participant_id;
  }

  /** Clone any template (factory preset or own) into a NEW non-builtin copy
   *  (records cloned_from_id server-side). Returns the new copy view. */
  async function clone(id: string): Promise<AgentTemplateView> {
    const w = await apiJson<AgentTemplateWire>(
      "POST",
      `/api/chat/agent-templates/${encodeURIComponent(id)}/clone`,
    );
    const view = wireToTemplate(w);
    templates.value = [...templates.value, view];
    return view;
  }

  /** Reset a "clone of a factory preset" copy back to its source content (the
   *  copy id is preserved). Returns the reset view. */
  async function reset(id: string): Promise<AgentTemplateView> {
    const w = await apiJson<AgentTemplateWire>(
      "POST",
      `/api/chat/agent-templates/${encodeURIComponent(id)}/reset`,
    );
    const view = wireToTemplate(w);
    templates.value = templates.value.map((tpl) => (tpl.id === id ? view : tpl));
    return view;
  }

  return {
    templates,
    loaded,
    fetchAll,
    create,
    update,
    remove,
    applyToConversation,
    clone,
    reset,
  };
});
